dlib/cuda/cuda_dlib.cu

```diff
@@ -2201,8 +2168,73 @@ namespace dlib
             invstds.set_size(src.num_samples());
             means = 0;
             invstds = 0;
-            launch_kernel(_cuda_layer_normalize, max_jobs(num, src.num_samples()), dest.device(), src.device(),
-                          means.device(), invstds.device(), gamma.device(), beta.device(), eps, src.num_samples(), num);
+            launch_kernel(_cuda_layer_normalize, max_jobs(src.k() * num, src.num_samples()), dest.device(), src.device(),
+                          means.device(), invstds.device(), gamma.device(), beta.device(), eps, src.num_samples(), src.k(), num);
+        }
+
+    // ----------------------------------------------------------------------------------------
+
+        __global__ void _cuda_layer_normalize_gradient(
+            float* out,
+            float* gg,
+            float* bg,
+            const float* s,
+            const float* gi,
+            const float* m,
+            const float* v,
+            const float* g,
+            float* dm,
+            float* dv,
+            float eps,
+            size_t ns,
+            size_t ks,
+            size_t num)
+        {
+            for (auto n : grid_stride_range_y(0, ns))
+            {
+                const auto ps = s + n * ks * num;
+                const auto pgi = gi + n * ks * num;
+                const float invstd_pow = -0.5 * std::pow(v[n], 3.0f);
+                float temp_dv = 0;
+                for (auto i : grid_stride_range(0, ks * num))
+                {
+                    const float x_hat = (ps[i] - m[n]) * v[n];
+                    atomicAdd(&bg[i / num], pgi[i]);
+                    atomicAdd(&gg[i / num], pgi[i] * x_hat);
```

Comment:
Not too happy about this. I will fix it tonight.

It makes the backward pass awfully slow.
In my experiment, a network trains at about 320 img/s, if I comment out these two lines, it trains at about 2560 img/s (similar rate as the batch norm)